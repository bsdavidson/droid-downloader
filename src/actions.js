export const DOWNLOAD_FILE = "DOWNLOAD_FILE";
export const DOWNLOAD_FOLDER = "DOWNLOAD_FOLDER";
export const REFRESH_DEVICES = "REFRESH_DEVICES";
export const REFRESH_DEVICE_FILES = "REFRESH_DEVICE_FILES";
export const SET_DEVICE = "SET_DEVICE";
export const SET_DEVICES = "SET_DEVICES";
export const SET_DEVICE_FILE = "SET_DEVICE_FILE";
export const SET_DEVICE_FILES = "SET_DEVICE_FILES";
export const SET_DEVICE_PATH = "SET_DEVICE_PATH";
export const SET_FILE_PREVIEW_IMAGE = "SET_FILE_PREVIEW_IMAGE";
export const SET_LOCAL_PATH = "SET_LOCAL_PATH";
export const TOGGLE_FILE_PREVIEW = "TOGGLE_FILE_PREVIEW";
export const TRAVERSE_DEVICE_PATH = "TRAVERSE_DEVICE_PATH";

export function downloadFile(name) {
  return {type: DOWNLOAD_FILE, name};
}

export function downloadFolder() {
  return {type: DOWNLOAD_FOLDER};
}

export function refreshDeviceFiles() {
  return {type: REFRESH_DEVICE_FILES};
}

export function refreshDevices() {
  return {type: REFRESH_DEVICES};
}

export function setDevice(device) {
  return {type: SET_DEVICE, device};
}

export function setDeviceFile(deviceFile) {
  return {type: SET_DEVICE_FILE, deviceFile};
}

export function setDeviceFiles(deviceFiles) {
  return {type: SET_DEVICE_FILES, deviceFiles};
}

export function setDevicePath(devicePath) {
  return {type: SET_DEVICE_PATH, devicePath};
}

export function setDevices(devices) {
  return {type: SET_DEVICES, devices};
}

export function setfilePreviewImage(filePreviewImage) {
  return {type: SET_FILE_PREVIEW_IMAGE, filePreviewImage};
}

export function setLocalPath(localPath) {
  return {type: SET_LOCAL_PATH, localPath};
}

export function toggleFilePreviewFullscreen() {
  return {type: TOGGLE_FILE_PREVIEW};
}

export function traverseDevicePath(dirName) {
  return {type: TRAVERSE_DEVICE_PATH, dirName};
}
