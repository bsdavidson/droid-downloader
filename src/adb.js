import {spawn} from "child_process";
import fs from "fs-extra";
import path from "path";

class ADBExitError extends Error {
  constructor(code, signal, stderr) {
    super(`ADB exited with code ${code} (signal: ${signal}) - ${stderr}`);
    this.code = code;
    this.signal = signal;
    this.stderr = stderr;
  }
}

class Device {
  constructor(serial, state, properties) {
    this.serial = serial;
    this.state = state;
    this.properties = properties;
  }

  static parseLine(line) {
    const fields = line.split(/ +/);
    return new Device(
      fields[0],
      fields[1],
      fields.slice(2).reduce((p, s) => {
        const [k, v] = s.split(":");
        p[k] = v; // eslint-disable-line no-param-reassign
        return p;
      }, {})
    );
  }
}

class DirLine {
  constructor(line, type, timestamp, name) {
    this.type = type;
    this.timestamp = timestamp;
    this.name = name;
    this.line = line;
  }

  static parseLine(line) {
    const parts = line
      .trim()
      .match(
        /^([^ ]+) +(?:[^ ]+ +){2,4}(\d{4}-\d{2}-\d{2} +\d{2}:\d{2}) +(.+)$/
      );
    if (!parts) {
      return null;
    }

    return new DirLine(
      line,
      parts[1].charAt(0) === "d" || parts[1].charAt(0) === "l"
        ? "directory"
        : "file",
      parts[2],
      parts[3].trim()
    );
  }
}

function shellQuote(value) {
  // (╯°□°）╯︵ ┻━┻
  return "'" + value.replace("'", `'"'"'`) + "'"; // eslint-disable-line prefer-template, quotes
}

export function getDevices() {
  return new Promise((resolve, reject) => {
    const process = spawn("/usr/local/bin/adb", ["devices", "-l"]);
    let stdoutBuffer = "";
    let stderrBuffer = "";

    process.stdout.on("data", data => {
      stdoutBuffer += data.toString();
    });

    process.stderr.on("data", data => {
      stderrBuffer += data.toString();
    });

    process.on("close", (code, signal) => {
      if (code) {
        reject(new ADBExitError(code, signal, stderrBuffer));
        return;
      }
      resolve(
        stdoutBuffer
          .trim()
          .split("\n")
          .slice(1)
          .map(Device.parseLine)
      );
    });
  });
}

export function getDirectoryList(deviceSerial, dirPath) {
  return new Promise(resolve => {
    const process = spawn("/usr/local/bin/adb", [
      "-s",
      deviceSerial,
      "shell",
      "ls",
      "-l",
      shellQuote(dirPath)
    ]);
    let stdoutBuffer = "";
    let stderrBuffer = "";

    process.stdout.on("data", data => {
      stdoutBuffer += data.toString();
    });

    process.stderr.on("data", data => {
      stderrBuffer += data.toString();
    });

    process.on("close", (code, signal) => {
      if (code) {
        // It is expected that ls will exit with an error code when it
        // encounters files or directories that it doesn't have permission
        // to access.
        console.warn(new ADBExitError(code, signal, stderrBuffer));
      }
      resolve(
        stdoutBuffer
          .trim()
          .split("\n")
          .map(DirLine.parseLine)
          .filter(d => d)
      );
    });
  });
}

export function getFileList(rootPath) {
  return new Promise((resolve, reject) => {
    let fileList = [];
    fs.readdir(rootPath, (err, names) => {
      if (err) {
        reject(err);
        return;
      }
      let remaining = names.length;
      function done(delta) {
        remaining -= delta;
        if (remaining === 0) {
          fileList.sort();
          resolve(fileList);
        }
      }
      names.forEach(name => {
        const filePath = path.join(rootPath, name);
        fs.lstat(filePath, (statErr, stats) => {
          if (statErr) {
            console.log(`error in lstat for ${filePath}: ${statErr}`);
          } else if (stats.isDirectory()) {
            getFileList(filePath).then(items => {
              fileList = fileList.concat(items);
              done(1);
            });
            return; // prevent done() from being called too soon
          } else if (stats.isFile()) {
            fileList.push(filePath);
          }
          done(1);
        });
      });
      done(0);
    });
  });
}

export function pull(deviceSerial, source, destination) {
  console.log(deviceSerial, source, destination);
  return new Promise((resolve, reject) => {
    const process = spawn("/usr/local/bin/adb", [
      "-s",
      deviceSerial,
      "pull",
      source,
      destination
    ]);
    let stdoutBuffer = "";
    let stderrBuffer = "";

    process.stdout.on("data", data => {
      stdoutBuffer += data.toString();
    });

    process.stderr.on("data", data => {
      stderrBuffer += data.toString();
    });

    process.on("close", (code, signal) => {
      if (code) {
        reject(new ADBExitError(code, signal, stderrBuffer));
        return;
      }
      resolve(stdoutBuffer.trim().split("\n"));
    });
  });
}
