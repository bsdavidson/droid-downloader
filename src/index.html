<!DOCTYPE html>
<html lang="en">

<head>
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no">
  <link href="./fontawesome/css/all.css" type="text/css" rel="stylesheet">
  <link href="./styles/index.scss" type="text/scss" rel="stylesheet">
</head>

<body>
  <div id="App"></div>

  <script>
    require('./index')
  </script>
</body>

</html>