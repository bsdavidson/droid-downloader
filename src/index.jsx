import React from "react";
import ReactDOM from "react-dom";
import {createStore, applyMiddleware} from "redux";
import thunk from "redux-thunk";
import {Provider} from "react-redux";
import {AppContainer} from "react-hot-loader";

import {reducer} from "./reducers";

const store = createStore(reducer, applyMiddleware(thunk));

const render = () => {
  const App = require("./app").default;
  ReactDOM.render(
    <Provider store={store}>
      <AppContainer>
        <App />
      </AppContainer>
    </Provider>,
    document.getElementById("App")
  );
};

render();
if (module.hot) {
  module.hot.accept(render);
}
