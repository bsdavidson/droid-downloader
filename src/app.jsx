import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {ipcRenderer} from "electron";

import {refreshDevices, setLocalPath} from "./actions";

import DeviceToolbar from "./device_toolbar";
import DeviceList from "./device_list";
import DirectoryList from "./directory_list";
import FileInfo from "./file_info";

export class App extends React.Component {
  static handleLocalPathSetClick() {
    ipcRenderer.send("open-file-dialog");
  }

  componentDidMount() {
    this.props.refreshDevices();

    ipcRenderer.on("selected-directory", (event, selectedPath) => {
      this.props.onLocalPathChange(selectedPath);
      // organizeFiles(selectedPath, selectedPath).catch(erroredFiles => {
      //   console.log("erroredFiles", erroredFiles);
      // });
    });
  }

  render() {
    return (
      <div className="app">
        <div className="app-device-list">
          <DeviceList />
        </div>
        <div className="app-directory-list">
          <DeviceToolbar />
          <DirectoryList />
          <FileInfo />
        </div>
      </div>
    );
  }
}

App.propTypes = {
  refreshDevices: PropTypes.func.isRequired,
  onLocalPathChange: PropTypes.func.isRequired
};

function mapDispatchToProps(dispatch) {
  return {
    refreshDevices: () => dispatch(refreshDevices()),
    onLocalPathChange: localPath => dispatch(setLocalPath(localPath))
  };
}

export default connect(
  null,
  mapDispatchToProps
)(App);
