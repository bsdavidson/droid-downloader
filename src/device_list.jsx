import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";

import {setDevice, refreshDevices} from "./actions";
import DroidPropTypes from "./prop_types";

export class DeviceList extends React.Component {
  render() {
    const {devices, device} = this.props;

    return (
      <div className="device-list">
        {devices.length > 0 ? (
          devices.map(d => (
            <div
              className={`device-list-item device-list-item-${
                d.serial === device ? "active" : "inactive"
              }`}
              key={d.serial}>
              <a
                href="#device"
                className="device-list-item-link"
                onClick={() => this.props.onDeviceClick(d.serial)}>
                <i className="device-list-item-icon fas fa-mobile-alt" />
                <span className="device-list-item-model">
                  {d.properties.model.replace(/_/g, " ")}
                </span>
              </a>
            </div>
          ))
        ) : (
          <span className="device-list-nodevices">No Devices Found</span>
        )}
        <div className="device-list-refresh">
          <a
            className="device-list-refresh-link"
            href="#refresh"
            onClick={this.props.onRefreshDevicesClick}>
            <i className="device-list-refresh-icon fas fa-sync-alt" />
            Refresh
          </a>
        </div>
      </div>
    );
  }
}

DeviceList.propTypes = {
  device: DroidPropTypes.device.isRequired,
  devices: DroidPropTypes.devices.isRequired,
  onDeviceClick: PropTypes.func.isRequired,
  onRefreshDevicesClick: PropTypes.func.isRequired
};

function mapStateToProps(state) {
  return {
    devices: state.devices,
    device: state.device
  };
}

function mapDispatchToProps(dispatch) {
  return {
    onDeviceClick: device => dispatch(setDevice(device)),
    onRefreshDevicesClick: () => dispatch(refreshDevices())
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(DeviceList);
