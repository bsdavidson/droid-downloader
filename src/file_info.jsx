import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";

import {downloadFile, toggleFilePreviewFullscreen} from "./actions";

export class FileInfo extends React.Component {
  constructor(props) {
    super(props);

    this.handleDownloadClick = this.handleDownloadClick.bind(this);
  }
  handleDownloadClick() {
    this.props.onDownloadClick(this.props.deviceFile.name);
  }
  render() {
    const {deviceFile, devicePath, filePreviewPath} = this.props;
    if (!deviceFile) {
      return null;
    }
    let previewImage;
    if (filePreviewPath) {
      previewImage = (
        <img
          className="file-info-preview-image"
          src={`file://${filePreviewPath}`}
          alt={deviceFile.name}
        />
      );
    } else {
      previewImage = (
        <p className="file-info-preview-none">No Preview Available</p>
      );
    }

    return (
      <div className="file-info">
        <div className="file-info-left">
          <div className="file-info-group file-info-path">
            <span className="file-info-label">Path:</span>
            <span className="file-info-value">
              {`/${devicePath.join("/")}`}
            </span>
          </div>
          <div className="file-info-group file-info-name">
            <span className="file-info-label">Filename:</span>
            <span className="file-info-value">{deviceFile.name}</span>
          </div>
          <div className="file-info-group file-info-size">
            <span className="file-info-label">File size:</span>
            <span className="file-info-value">{deviceFile.size} bytes</span>
          </div>
          <div className="file-info-group file-info-timestamp">
            <span className="file-info-label">Modified:</span>
            <span className="file-info-value">{deviceFile.timestamp}</span>
          </div>
          <div className="file-info-download">
            <button
              className="file-info-download-link"
              onClick={this.handleDownloadClick}>
              <i className="fas fa-file-download file-info-download-icon" />
              Download this file
            </button>
          </div>
        </div>
        <div className="file-info-right">
          <div className="file-info-preview">
            <a
              href="#previewFull"
              onClick={this.props.onPreviewFullScreenClick}>
              {previewImage}
            </a>
          </div>
        </div>
      </div>
    );
  }
}

FileInfo.defaultProps = {
  deviceFile: null,
  filePreviewPath: null
};

FileInfo.propTypes = {
  deviceFile: PropTypes.objectOf(PropTypes.string),
  devicePath: PropTypes.arrayOf(PropTypes.string).isRequired,
  filePreviewPath: PropTypes.string,
  onDownloadClick: PropTypes.func.isRequired,
  onPreviewFullScreenClick: PropTypes.func.isRequired
};

function mapStateToProps(state) {
  return {
    deviceFile: state.deviceFile,
    devicePath: state.devicePath,
    filePreviewPath: state.filePreviewPath
  };
}

function mapDispatchToProps(dispatch) {
  return {
    onDownloadClick: fileName => dispatch(downloadFile(fileName)),
    onPreviewFullScreenClick: () => {
      dispatch(toggleFilePreviewFullscreen());
    }
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(FileInfo);
