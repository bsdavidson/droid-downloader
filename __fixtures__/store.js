import {Device, DirLine} from "../src/adb";

export const device = "ffffbeefface";

export const deviceFiles = [
  new DirLine(
    "directory",
    "2018-01-02 09:45",
    "folder_1",
    "drwxr-xr-x",
    "3001"
  ),
  new DirLine("file", "2017-09-20 21:22", "file_1", "-rw-r--r--", "42100"),
  new DirLine("file", "2017-10-17 01:45", "file_2", "-rw-r--r--", "3125995")
];

export const deviceFile = deviceFiles[1];

export const devicePath = ["storage", "emulated", "0", "DCIM"];

export const devices = [
  new Device("ffff1337face", "device", {
    model: "SM_G955U"
  }),
  new Device("ffffbeefface", "device", {
    model: "SM_G999V"
  }),
  new Device("ffffbadbeef", "device", {
    model: "SM_G910V"
  })
];

export const localPath = "/tmp";

export const testImage =
  "data:image/png;base64, iVBORw0KGgoAAAANSUhEUgAAADIAAAAyCAMAAAAp4XiDAAAARVBMVEX/////AAAAAAAAAP8A/wD39/c2yQA7Ozu8vP8ACvUA8wz/4+P/9PT09PTj4//09P/j4+P0//Tj/+Po+eP3/fTp6en4+P+6GBu+AAAAUklEQVRIie3RNw6AMBQE0U8wOaf7H5XKsjtGFqJA+/qpxkw+0mfMoCQ92SvmCMlWMutPk46Jkqll5igpGCXe1TAm6c6acSFZcmZU4jnmtavy4AblqhLOouOKVQAAAABJRU5ErkJggg==";
