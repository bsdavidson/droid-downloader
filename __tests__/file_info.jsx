import React from "react";
import {shallow} from "enzyme";
import renderer from "react-test-renderer";

import {FileInfo} from "../src/file_info";
import {deviceFile, devicePath} from "../__fixtures__/store";

test("<FileInfo /> should render an infoBox", () => {
  const wrapper = renderer
    .create(
      <FileInfo
        devicePath={devicePath}
        deviceFile={deviceFile}
        onDownloadClick={() => {}}
      />
    )
    .toJSON();
  expect(wrapper).toMatchSnapshot();
});

test("<FileInfo /> should not render when no file is selected", () => {
  const wrapper = renderer
    .create(
      <FileInfo
        devicePath={devicePath}
        deviceFile={null}
        onDownloadClick={() => {}}
      />
    )
    .toJSON();
  expect(wrapper).toMatchSnapshot();
});

test("<FileInfo /> should call onDownloadClick when download button is clicked", () => {
  const onDownloadClick = jest.fn();

  const wrapper = shallow(
    <FileInfo
      devicePath={devicePath}
      deviceFile={deviceFile}
      onDownloadClick={onDownloadClick}
    />
  );

  expect(onDownloadClick.mock.calls.length).toBe(0);
  wrapper.find(".file-info-download-link").simulate("click");
  expect(onDownloadClick.mock.calls.length).toBe(1);
  expect(onDownloadClick.mock.calls[0]).toEqual([deviceFile.name]);
});
