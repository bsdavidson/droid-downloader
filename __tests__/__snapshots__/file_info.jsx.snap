// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FileInfo /> should not render when no file is selected 1`] = `null`;

exports[`<FileInfo /> should render an infoBox 1`] = `
<div
  className="file-info"
>
  <div
    className="file-info-left"
  >
    <div
      className="file-info-group file-info-path"
    >
      <span
        className="file-info-label"
      >
        Path:
      </span>
      <span
        className="file-info-value"
      >
        /storage/emulated/0/DCIM
      </span>
    </div>
    <div
      className="file-info-group file-info-name"
    >
      <span
        className="file-info-label"
      >
        Filename:
      </span>
      <span
        className="file-info-value"
      >
        file_1
      </span>
    </div>
    <div
      className="file-info-group file-info-size"
    >
      <span
        className="file-info-label"
      >
        File size:
      </span>
      <span
        className="file-info-value"
      >
        42100
         bytes
      </span>
    </div>
    <div
      className="file-info-group file-info-timestamp"
    >
      <span
        className="file-info-label"
      >
        Modified:
      </span>
      <span
        className="file-info-value"
      >
        2017-09-20 21:22
      </span>
    </div>
    <div
      className="file-info-download"
    >
      <button
        className="file-info-download-link"
        onClick={[Function]}
      >
        <i
          className="fas fa-file-download file-info-download-icon"
        />
        Download this file
      </button>
    </div>
  </div>
  <div
    className="file-info-right"
  >
    <div
      className="file-info-preview"
    >
      <a
        href="#previewFull"
      >
        <p
          className="file-info-preview-none"
        >
          No Preview Available
        </p>
      </a>
    </div>
  </div>
</div>
`;
